src/Psl/DataStructure/QueueInterface.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\DataStructure;
+
+use Psl;
+use Countable;
+
+/**
+ * An interface representing a queue data structure ( FIFO ).
+ *
+ * @template T
+ *
+ * @see https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)
+ */
+interface QueueInterface extends Countable
```

Comment:
How would you use the queue?
I assume in most cases you want to iterate over the queue.
However, this leaves you with this implementation:

```php
while ($res = $queue->pull() !== null) {
}
```

I find this quite awful :)
Can we make those classes `Iterable`? 
That way you can also use the `Iter\` functions on a queue.


(same remarks for all those interfaces)

Reply:
I thought about implementing iterator or iterator aggregate, but not sure which would be better?

implementing iterator means that the queue might change within the loop, resulting probably in a corupted queue, implementing iterator aggregate, means values enqued within the loop, won't be iteratored over.

which one do you think makes more sense?

Reply:
That is a good question. Did not think about that tbh ...
Maybe you can think about the use-cases : would it make sense to be able to queue something whilst consuming the items inside the queue?
For example the stack : if you add something, it will be picked up in the next iteration.


You could also take a look how the SPL variants work and determine if it makes sense:

- https://www.php.net/manual/en/class.splstack.php
- https://www.php.net/manual/en/class.splpriorityqueue.php
- https://www.php.net/manual/en/class.splqueue.php


I do like the idea of being able to add items while iterating, but I don't know if it will be used in the real world.
Places where I used priority queues was e.g. inside event listeners - which are configured before consumed.
That specific implementation did not clear the queue whilst iterating.

On the other hand : the SPL provides `Iterator` and not `IteratorAggregate`.

Reply:
I can't think of a case where I had to push items into a queue while iterating over it either. and I think generally enquing and consuming should be done seperatly, therefor we should implement `IteratorAggregate`

Reply:
Ah, another thing, should we remove items for the queue when iteratoring over them?

Reply:
currently, SPL doesn't ... while consuming a queue or a stack, i would assume that items are removed ... this is quite weird.

https://3v4l.org/WL8os

Reply:
WDYT about adding `empty(): bool` instead? 🤔 

I think when iterating over a  queue, people will have differente expectations ( can I enqueue while consuming? will i be consuming the nodes i pushed? will they be removed from the queue? )

while the following is pretty clear:

```php
while(!$queue->empty()) {
  $node = $queue->dequeue();

  // use the node. 
  if (something) {
    $queue->enqueue(something); // will be consumed in this loop
  }
}
```

Reply:
let's merge as is now, and continue discussing this in #83 for later.